### ISSUE TITLE
show more information in server list

### ISSUE THREAD

**Issue by InflamedSebi** [PufferPanel/PufferPanel#924]
Having an owner/responsible displayed on every server in the list including the used template would help when cleaning up or finding the right server.

**Comment by nepcore** [comment]
A server doesn't know what template it was created from, that information is not retained anywhere  
Also, a template can change, as servers are not automatically changed when a template changes, displaying what template was used would likely cause confusion for some people as two servers created from the same template can behave in completely different ways as they could have been created at different times with significant changes to the template in between  
iirc servers do keep the templates type property (e.g. `minecraft-java`), we could display that on the server list if we can find a way to do so without cluttering it up too much, but I'd need @LordRalex to confirm if I remember that correctly

PufferPanel also has no clue who is responsible for a server as there is no concept of ownership to servers, there are just users that have some amount of permission to control certain aspects of a server, though i could imagine that it might be possible to add a way of retrieving a servers assigned users without leaving the list page e.g. by having an info icon next to the status indicator that when hovered or clicked could display some additional info (like users assigned to the server), not sure how well that would work though

**Comment by InflamedSebi** [comment]
I'm fine with any hint on the template, so `minecraft-java` is good enough.
For the owner/responsible thing. I started adding a username to the servers title, which does work well, as long as the owners don't change.

However, now that I think of it again, having a small custom text annotation (which can be edited somewhere in the Server GUI) would probably be superior.
Visually in the server list maybe  something like the popups in forms:
![amtrak](https://user-images.githubusercontent.com/1872197/106362098-52411680-6321-11eb-8e13-b58c91a3a7d2.png)

So I can put the creation date, the person who wanted it and maybe a note. While other users of pufferpanel might have different information in there, since it is free text.

**Comment by nepcore** [comment]
A generic notes field would to me as a user feel like the creator didn't want to think about how to properly solve the problem, so they pawned it off onto the user to do so themselves with halfassed tooling  
I think it would be much better to figure out what information users actually need on the listing and just display that  
That also means that we should look into how important what part of the information that needs to make it onto the list is, so we can make sure that we display the most important info (where possible)

**Comment by InflamedSebi** [comment]
Only giving strict predefined information takes a lot of flexibility from the user and the software, so it tends to feels unflexible. Also different hosted services would probably profit, by having different information in the overview. But a agree, a simple free text popup, does feel a little lazy.

What do you think about using the same mechanics like the templates do? I mean the token/text replacing thing like `${rootDir}` or `${port}` and so on.
So instead of neither being hard coded nor a lazy note text this could become a part of the templates since further information may heavily depend on the template used.

The current one would probably look like this:
```
{
  "display": {
    "header": "${serverName}",
    "subtext": "${nodeFqdn}:${ServerPort} @ ${nodeName}"
  }
}
```
This would require a few more predefined placeholders than just `${rootDir}` but the more powerful thing would be having it lookup actual defined variables for this service, so an ark server might look like this:
```
{
  "display": {
    "header": "${serverName} (${sessionName})",
    "subtext": "${nodeFqdn}:${ServerPort} @ ${nodeName}",
    "popup": "Friends only: ${exclusive}\nMap: ${mapName}\nSlots: ${maxPlayers}\n\n${notes}"
  }
}
```
Everyone can adjust this to make their overview as plain or complex as they want. Don't need the node name, since you only run one anyway? Leave it out or change it to something else.
On the backend side you can always just support new placeholders or offer different display options then just "header", "subtext" and "popup". Maybe an "accordion" or a "slider"? Found a good was to display all owners? add an `"ownerList": true`.
Templates wouldn't need to adjust (even current ones would work), since all of this is optional and if there is no "display" section in the template, or it does not parse correctly, you can fallback to defaults.

I like things where you can go crazy if you want, but you don't need to at all.
What do you think?

**Comment by nepcore** [comment]
`display` is already used in templates, so that's not an option, but that's just naming  
Also, I don't think it's a good idea to give templates control over _every_ aspect of what is displayed on the server list as that is bound to lead to loads of different approaches making the whole thing messy as all hell, at least the list entries header should be fixed

Allowing templates to takes some control over information on the server list does make sense to me, however there are some bigger issues with doing that
The API would suddenly need to run to each servers node to collect data (currently it just goes off of what the panel knows about) which is 10 HTTP(S) requests per list call  
Allowing variable usage would additionally mean for each server the node then has to look up variables and predefs to put those in  
And if we were to also make things like the node info part of that we'd then need the API to define and apply its own predefs as well  
That much logic _per server_ to me feels like a really bad idea for something that is bound to be used as much as the server list

One could argue you only need to update it whenever any variable changes and save a compiled version in the panel db, but that then falls short when the user makes use of the daemon API to update variables without the panel being in the loop, so the panel can't know the correct state anymore, which means we're back to 10 node API calls per list page again

So while I do like the idea of a template being able to define some data shown in the server list I don't think we can really do that without the server list suddenly becoming _a lot_ slower and getting worse with every additional server

**Comment by disconsented** [comment]
Well as an alternative how about returning a (possibility sanitized version of) server struct and then rendering that out?

Still one call, generally should be able to handle most uses cases. The only thing that would be needed is some logic to adjust how that is then rendered out.

**Comment by InflamedSebi** [comment]
I would also disagree on caching/storing pre-rendered data. This would totally end up out of sync, like it already does with the port. I agree to better create it just in time, if there is a performant solution.
It also doesn't need to be fully customizable ofc, there is probably a good balance to be found.

If I get you right, disconsented, We need one endpoint which is called once per server, that would deliver all required information for the rendering in one go, like the placeholders with substitutions plus the part of the template that does define the display. So all the rendering would be client side.
Or may it be better to pre-build the rendering server side and have the endpoint deliver display ready data? The Endpoint would be less useful for other usages, though. Probably not best practice ...

**Comment by disconsented** [comment]
It'd be once per node but yes in essence.

It may be worth investigating how Vue handles plugins to facilitate this.

**Comment by InflamedSebi** [comment]
Another important stat to list in in overview might be disk space.
Current used space for this server similar to `du -sh . `
And available space of the containing partition like `df -hP . | awk 'NR==2 {print $4}'`
So if servers run on different nodes or the available space for them is individual (f.e. zfs datasets) you can tell which is about to reach its limit.
As far as i can tell there is no place in the GUI which does tell you that information at all.

**Comment by disconsented** [comment]
That's a topic for a separate discussion

**Comment by InflamedSebi** [comment]
Yeah it might be, but it is related, since it is one of the informations I would like to see in the server list :)  and I don't want to open multiple issues on the the same topic.

**Comment by disconsented** [comment]
So I see this as two separate problems:

1) Extensible * for the front end @nepcore could weigh in on this, my vue is very rusty.
2) Disk 'management'
 2a) Keeping track of usage can be expensive
 2b) Integration into limiting disk resources for SFTP (?) and support for docker (since there isn't support for this *nix otherwise afaik)

**Comment by InflamedSebi** [comment]
Oh, this was not a request for disk management, just for plain information on disk usage similar to the Ram/cpu load. So I can stay informed without sshing into the server itself or a separate monitoring solution.

The disk management was a topic lately on discord and can be done outside of puffer using zfs as filesystem (quite easy to setup limits for folders) or if you dont want to use zfs by formatting a file as a filesystem and mounting that via dev/loop to the correct location, which should work for docker and SFTP just fine.

I totally get monitoring disc space might be expensive, but I personally don't need live data. The current usage at the time the overview gets loaded, would suffice for me, so it could just be a single data point in the same endpoint that does deliver the information for the server list.
